### ISSUE TITLE
TypeError: path must be a string or Buffer

### ISSUE THREAD

**Issue by ismaelbej** [trufflesuite/truffle-core#88]
## Issue

While trying to run a test in solidity it throws the error message: "TypeError: path must be a string or Buffer".

## Steps to Reproduce

File: TestWakeUp.sol

```
pragma solidity ^0.4.19;
import "truffle/Assert.sol";
constract TestWakeup {
  function testSleeping() public {
    Assert.equal(1, 1, "I should be sleeping");
  }
}
```

After triagging the issue is in the `orderTestContractFunctions` , at this line https://github.com/trufflesuite/truffle-core/blob/b3ad375993ec42bc622c7674258edc7614944482/lib/test.js#L211

```
async.each(contracts, function(contract, finished) {
        var file = file_hash[contract.contract_name];
```

Here file is null. 

My issue was I've named my contract with the wrong caption "TestWakeUp" vs "TestWakeup".

## Expected Behavior

A nice error message indicating the test has an invalid name.

## Actual Results

The test fails to compile and run.

```
    at Object.fs.readFile (fs.js:358:11)
    at Object.ordered_abi (C:\Users\\AppData\Roaming\npm\node_modules\truffle\build\cli.bundled.js:203860:8)
    at C:\Users\nomnom\AppData\Roaming\npm\node_modules\truffle\build\cli.bundled.js:318977:23
    at C:\Users\nomnom\AppData\Roaming\npm\node_modules\truffle\build\cli.bundled.js:31984:20
    at Object.async.forEachOf.async.eachOf (C:\Users\nomnom\AppData\Roaming\npm\node_modules\truffle\build\cli.bundled.js:32036:13)
    at Object.async.forEach.async.each (C:\Users\nomnom\AppData\Roaming\npm\node_modules\truffle\build\cli.bundled.js:32012:22)
    at C:\Users\nomnom\AppData\Roaming\npm\node_modules\truffle\build\cli.bundled.js:318972:13
    at new Promise (<anonymous>)
    at Object.orderTestContractFunctions (C:\Users\nomnom\AppData\Roaming\npm\node_modules\truffle\build\cli.bundled.js:318964:12)
    at C:\Users\nomnom\AppData\Roaming\npm\node_modules\truffle\build\cli.bundled.js:318879:19
```

## Environment

* Operating System:  Windows 7 x64
* Truffle version: Truffle v4.0.4 (core: 4.0.4)
* Ethereum client: Ganache CLI v6.0.3 (ganache-core: 2.0.2)
* node version: node v8.9.0
* npm version: npm v5.6.0

**Comment by cgewecke** [comment]
@ismaelbej Thanks for such a clear description and diagnosis of the problem. 

This should be resolved (or made different anyway) by the movement of ABI ordering logic to `truffle-compile`. Should be closable with release `4.0.6`

**Comment by sabhiram** [comment]
Maybe this is a typo - but in your contract definition you have `constract TestWakeup {` instead of `contract TestWakeup {`

**Comment by ismaelbej** [comment]
@sabhiram  You are right, also it is in the report. It is very easy to make error and the message is not helpful to find the problem. I reported the error message so if someone else gets bitten by the same bug, there will be an issue where he can look for a solution.

**Comment by gnidan** [comment]
This should be fixed in v4.0.6, since we removed that code. Let us know if the problem persists!
